src/core/reactor_backend.cc

```diff
@@ -1559,16 +1559,16 @@ class reactor_backend_uring final : public reactor_backend {
     virtual void wait_and_process_events(const sigset_t* active_sigmask) override {
         _smp_wakeup_completion.maybe_rearm(*this);
         _hrtimer_completion.maybe_rearm(*this);
-        ::io_uring_submit(&_uring);
         bool did_work = false;
         did_work |= _preempt_io_context.service_preempting_io();
         did_work |= std::exchange(_did_work_while_getting_sqe, false);
         if (did_work) {
+            ::io_uring_submit(&_uring);
             return;
         }
         struct ::io_uring_cqe* cqe = nullptr;
         sigset_t sigs = *active_sigmask; // io_uring_wait_cqes() wants non-const
-        auto r = ::io_uring_wait_cqes(&_uring, &cqe, 1, nullptr, &sigs);
+        auto r = ::io_uring_submit_and_wait_timeout(&_uring, &cqe, 1, nullptr, &sigs);
```

Comment:
This returns -ETIME on timeout, but you don't handle it. How did it work?

Reply:
`-ETIME` is returned only if a `ts` is specified. see https://github.com/axboe/liburing/blob/45ed5c68150e1438b6145e3390af2c3b17e90f4a/src/queue.c#L349 , where `__io_uring_get_cqe()` gets called. and `__io_uring_get_cqe()` passes a `get_data` instance to [`_io_uring_get_cqe()`](https://github.com/axboe/liburing/blob/45ed5c68150e1438b6145e3390af2c3b17e90f4a/src/queue.c#L62) without setting its `has_ts`, see https://github.com/axboe/liburing/blob/45ed5c68150e1438b6145e3390af2c3b17e90f4a/src/queue.c#L135 . and `-ETIME` is returned only if `data->has_ts` is set. see https://github.com/axboe/liburing/blob/45ed5c68150e1438b6145e3390af2c3b17e90f4a/src/queue.c#L107 . and this function calls into `io_uring_enter2` syscall, which is an alias of `io_uring_enter`. but this syscall does not return `-ETIME`. see https://github.com/axboe/liburing/blob/master/man/io_uring_enter.2